lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+      mlir::Value resultMux = muxOp.getResult();
+      mlir::Value select = muxOp.getSelectOperand();
+      mlir::Value condition = condBranchOp.getConditionOperand();
+      mlir::Value branchIn = condBranchOp.getDataOperand();
+      mlir::Value trueResult = condBranchOp.getTrueResult();
+      mlir::Value falseResult = condBranchOp.getFalseResult();
+      mlir::ValueRange l = {second, second};
+
+      if (resultMux == branchIn) {
+        if (trueResult == first) {
+          rewriter.setInsertionPoint(muxOp);
+          rewriter.create<handshake::SinkOp>(muxOp->getLoc(), select);
+          rewriter.replaceOp(muxOp, second);
+          rewriter.replaceAllUsesWith(falseResult, second);
+        } else if (trueResult == second) {
+          rewriter.setInsertionPoint(muxOp);
+          rewriter.create<handshake::SinkOp>(muxOp->getLoc(), select);
+          rewriter.replaceOp(muxOp, first);
+          rewriter.replaceAllUsesWith(falseResult, first);
+        } else if (falseResult == first) {
+          rewriter.setInsertionPoint(muxOp);
+          rewriter.create<handshake::SinkOp>(muxOp->getLoc(), select);
+          rewriter.replaceOp(muxOp, second);
+          rewriter.replaceAllUsesWith(trueResult, second);
+        } else if (falseResult == second) {
+          rewriter.setInsertionPoint(muxOp);
+          rewriter.create<handshake::SinkOp>(muxOp->getLoc(), select);
+          rewriter.replaceOp(muxOp, first);
+          rewriter.replaceAllUsesWith(trueResult, first);
+        } else {
+          return failure();
+        }
+        rewriter.setInsertionPoint(condBranchOp);
+        rewriter.create<handshake::SinkOp>(condBranchOp->getLoc(), condition);
+        rewriter.eraseOp(condBranchOp);
+        return success();
+      }
+    }
+    return failure();
+  }
+};
+
+struct RemoveMergeBranchLoopPairs
+    : public OpRewritePattern<handshake::MergeOp> {
+  using OpRewritePattern<handshake::MergeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::MergeOp mergeOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner = nullptr;
+    auto it = mergeOp->getUsers().begin();
+    if (it != mergeOp->getUsers().end()) {
+      useOwner = *it;
+    }
+    if (isa_and_nonnull<handshake::ConditionalBranchOp>(useOwner)) {
+      handshake::ConditionalBranchOp condBranchOp =
+          mlir::cast<handshake::ConditionalBranchOp>(useOwner);
+
+      mlir::OperandRange dataOperands = mergeOp.getDataOperands();
+      mlir::Value first = (dataOperands[0]);
+      mlir::Value second = (dataOperands[1]);
+      mlir::Value resultMerge = mergeOp.getResult();
+      mlir::Value condition = condBranchOp.getConditionOperand();
+      mlir::Value branchIn = condBranchOp.getDataOperand();
+      mlir::Value trueResult = condBranchOp.getTrueResult();
+      mlir::Value falseResult = condBranchOp.getFalseResult();
+
+      if (resultMerge == branchIn) {
```

Comment:
Same comment as in the previous pattern applies about avoid code repetition here.